### PR TITLE
When refreshing after a reboot, force lock allocation

### LINKED ISSUE #2900: podman rm -a shows "No such file or directory"
<!--
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST

**NOTE** A large number of issues reported against Podman are often found to already be fixed
in more current versions of the project.  Before reporting an issue, please verify the
version you are running with `podman version` and compare it to the latest release
documented on the top of Podman's [README.md](../README.md).  If they differ, please
update your version of Podman to the latest possible and retry your command before creating
an issue.

If you are filing a bug against `podman build`, please instead file a bug
against Buildah (https://github.com/containers/buildah/issues). Podman build
executes Buildah to perform container builds, and as such the Buildah
maintainers are best equipped to handle these bugs.
-->

**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug


**Description**

When running podman rm -a, an error message: "No such file or directory is shown"

**Steps to reproduce the issue:**

1. podman rm -a

**Describe the results you received:**

```
$ podman ps                                                                                                                                                                                                                                                                                                                                                                                  
CONTAINER ID  IMAGE  COMMAND  CREATED  STATUS  PORTS  NAMES                                                                                                                                                                                                                                                                                                                                  
 ```
    
```                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
$ podman ps -a                                                                                                                                                                                                                                                                                                                                                                               
CONTAINER ID  IMAGE                              COMMAND               CREATED            STATUS                          PORTS                 NAMES                                                                                                                                                                                                                                       
d28ccf5e7ab7  localhost/jorget/pbench-fio:v1.0   /bin/bash             2 minutes ago      Exited (0) 2 minutes ago                              elated_nobel                                                                                                                                                                                                                                
a2a771643f10  localhost/jorget/pbench-fio:v1.0   /opt/pbench-fio-t...  3 minutes ago      Created                                               objective_pare                                                                                                                                                                                                                              
84140756a6fb  localhost/jorget/pbench-fio:v1.0   /bin/bash             4 minutes ago      Exited (127) 3 minutes ago                            youthful_mcclintock                                                                                                                                                                                                                         
92bec0ca4591  localhost/jorget/pbench-fio:v1.0   /bin/bash             4 minutes ago      Exited (0) 4 minutes ago                              elegant_chatelet
b62b463b8865  localhost/jorget/pbench-fio:v1.0   /bin/bash             21 minutes ago     Exited (0) 7 minutes ago        0.0.0.0:2323->22/tcp  recursing_rhodes
7690f5398d3d  localhost/jorget/pbench-fio:v1.0   /bin/bash             22 minutes ago     Exited (137) 21 minutes ago     0.0.0.0:2323->22/tcp  vigilant_mendel
faa9013e2340  localhost/jorget/pbench-fio:v1.0   /usr/sbin/init        About an hour ago  Exited (137) About an hour ago  0.0.0.0:2323->22/tcp  nostalgic_antonelli
e30a57fa2823  localhost/jorget/pbench-fio:v1.0   /usr/sbin/init        About an hour ago  Exited (137) About an hour ago  0.0.0.0:2323->22/tcp  jovial_pike
97e2404083fd  localhost/jorget/pbench-fio:v1.0   /usr/sbin/init        About an hour ago  Created                                               inspiring_kepler
903b36684896  localhost/jorget/pbench-fio:v1.0   /usr/sbin/init        About an hour ago  Exited (137) About an hour ago  0.0.0.0:2323->22/tcp  fervent_meitner
715e64988c56  localhost/jorget/pbench-fio:v1.0   /usr/sbin/init        About an hour ago  Exited (137) About an hour ago  0.0.0.0:2323->22/tcp  clever_fermi
5e3f3ee7ee94  localhost/jorget/pbench-fio:v1.0   /usr/sbin/init        About an hour ago  Exited (137) About an hour ago  0.0.0.0:2323->22/tcp  crazy_jennings
ac45d9e6a55e  docker.io/library/elasticsearch:5  /docker-entrypoin...  8 weeks ago        Created                                               blissful_golick
b2767dbf6c6b  docker.io/library/elasticsearch:5  /docker-entrypoin...  4 months ago       Created                                               romantic_mayer
7336ba5ca27a  docker.io/pihole/pihole:latest     /s6-init              4 months ago       Created                                               elegant_knuth
```

```
$ podman --log-level=debug rm -a
INFO[0000] running as rootless
DEBU[0000] Initializing boltdb state at /home/jtudelag/.local/share/containers/storage/libpod/bolt_state.db
DEBU[0000] Overriding run root "/run/user/1000" with "/run/user/1000/run" from database
DEBU[0000] Using graph driver vfs
DEBU[0000] Using graph root /home/jtudelag/.local/share/containers/storage
DEBU[0000] Using run root /run/user/1000/run
DEBU[0000] Using static dir /home/jtudelag/.local/share/containers/storage/libpod
DEBU[0000] Using tmp dir /run/user/1000/libpod/tmp
DEBU[0000] Using volume path /home/jtudelag/.local/share/containers/storage/volumes
DEBU[0000] Set libpod namespace to ""
DEBU[0000] Not configuring container store
INFO[0000] running as rootless
WARN[0000] The configuration is using `runtime_path`, which is deprecated and will be removed in future.  Please use `runtimes` and `runtime`
WARN[0000] If you are using both `runtime_path` and `runtime`, the configuration from `runtime_path` is used
DEBU[0000] Initializing boltdb state at /home/jtudelag/.local/share/containers/storage/libpod/bolt_state.db
DEBU[0000] Overriding run root "/run/user/1000" with "/run/user/1000/run" from database
DEBU[0000] Using graph driver vfs
DEBU[0000] Using graph root /home/jtudelag/.local/share/containers/storage
DEBU[0000] Using run root /run/user/1000/run
DEBU[0000] Using static dir /home/jtudelag/.local/share/containers/storage/libpod
DEBU[0000] Using tmp dir /run/user/1000/libpod/tmp
DEBU[0000] Using volume path /home/jtudelag/.local/share/containers/storage/volumes
DEBU[0000] Set libpod namespace to ""
DEBU[0000] [graphdriver] trying provided driver "vfs"
DEBU[0000] Setting maximum workers to 16
DEBU[0000] Cleaning up container 5e3f3ee7ee945b3d628639c1515d52abf84bbbec22855776857cb9f2765e5e33
DEBU[0000] Network is already cleaned up, skipping...
DEBU[0000] Storage is already unmounted, skipping...
DEBU[0000] Storage is already unmounted, skipping...
5e3f3ee7ee945b3d628639c1515d52abf84bbbec22855776857cb9f2765e5e33
INFO[0000] running as rootless
WARN[0000] The configuration is using `runtime_path`, which is deprecated and will be removed in future.  Please use `runtimes` and `runtime`
WARN[0000] If you are using both `runtime_path` and `runtime`, the configuration from `runtime_path` is used
DEBU[0000] Initializing boltdb state at /home/jtudelag/.local/share/containers/storage/libpod/bolt_state.db
DEBU[0000] Overriding run root "/run/user/1000" with "/run/user/1000/run" from database
DEBU[0000] Using graph driver vfs
DEBU[0000] Using graph root /home/jtudelag/.local/share/containers/storage
DEBU[0000] Using run root /run/user/1000/run
DEBU[0000] Using static dir /home/jtudelag/.local/share/containers/storage/libpod
DEBU[0000] Using tmp dir /run/user/1000/libpod/tmp
DEBU[0000] Using volume path /home/jtudelag/.local/share/containers/storage/volumes
DEBU[0000] Set libpod namespace to ""
DEBU[0000] [graphdriver] trying provided driver "vfs"
DEBU[0000] Setting maximum workers to 16
DEBU[0000] Cleaning up container 715e64988c56b837da5af52f3f7dbbfac936dbb4404c28ae094bddd7d60a8667
DEBU[0000] Network is already cleaned up, skipping...
DEBU[0000] Storage is already unmounted, skipping...
DEBU[0000] Storage is already unmounted, skipping...
715e64988c56b837da5af52f3f7dbbfac936dbb4404c28ae094bddd7d60a8667
INFO[0000] running as rootless
WARN[0000] The configuration is using `runtime_path`, which is deprecated and will be removed in future.  Please use `runtimes` and `runtime`
WARN[0000] If you are using both `runtime_path` and `runtime`, the configuration from `runtime_path` is used
DEBU[0000] Initializing boltdb state at /home/jtudelag/.local/share/containers/storage/libpod/bolt_state.db
DEBU[0000] Overriding run root "/run/user/1000" with "/run/user/1000/run" from database
DEBU[0000] Using graph driver vfs
DEBU[0000] Using graph root /home/jtudelag/.local/share/containers/storage
DEBU[0000] Using run root /run/user/1000/run
DEBU[0000] Using static dir /home/jtudelag/.local/share/containers/storage/libpod
DEBU[0000] Using tmp dir /run/user/1000/libpod/tmp
DEBU[0000] Using volume path /home/jtudelag/.local/share/containers/storage/volumes
DEBU[0000] Set libpod namespace to ""
DEBU[0000] [graphdriver] trying provided driver "vfs"
DEBU[0000] Setting maximum workers to 16
DEBU[0000] Cleaning up container 7336ba5ca27acf31ec13286ac499e7d716c8d3637d54e3e65c749a2568ada7ac
DEBU[0000] Network is already cleaned up, skipping...
DEBU[0000] Storage is already unmounted, skipping...
DEBU[0000] Storage is already unmounted, skipping...
ERRO[0000] no such file or directory
```

**Describe the results you expected:**


**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
podman version
Version:            1.2.0
RemoteAPI Version:  1
Go Version:         go1.11.5
OS/Arch:            linux/amd64
```


**Output of `podman info --debug`:**
```
$ podman info --debug
debug:
  compiler: gc
  git commit: ""
  go version: go1.11.5
  podman version: 1.2.0
host:
  BuildahVersion: 1.7.2
  Conmon:
    package: podman-1.2.0-2.git3bd528e.fc29.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.12.0-dev, commit: d88bb0e63cb70f9787a8e410716924f380af361f'
  Distribution:
    distribution: fedora
    version: "29"
  MemFree: 3411947520
  MemTotal: 20434337792
  OCIRuntime:
    package: runc-1.0.0-68.dev.git6635b4f.fc29.x86_64
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc6+dev
      commit: ef9132178ccc3d2775d4fb51f1e431f30cac1398-dirty
      spec: 1.0.1-dev
  SwapFree: 10287050752
  SwapTotal: 10292817920
  arch: amd64
  cpus: 4
  hostname: jtudelag-t460s
  kernel: 5.0.5-200.fc29.x86_64
  os: linux
  rootless: true
  uptime: 42h 43m 35.56s (Approximately 1.75 days)
insecure registries:
  registries:
  - localhost:5000
store:
  ConfigFile: /home/jtudelag/.config/containers/storage.conf
  ContainerStore:
    number: 6
  GraphDriverName: vfs
  GraphOptions: null
  GraphRoot: /home/jtudelag/.local/share/containers/storage
  GraphStatus: {}
  ImageStore:
    number: 47
  RunRoot: /run/user/1000/run
  VolumePath: /home/jtudelag/.local/share/containers/storage/volumes

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by mheon**
After a reboot, when we refresh Podman's state, we retrieved the lock from the fresh SHM instance, but we did not mark it as allocated to prevent it being handed out to other containers and pods.

Provide a method for marking locks as in-use, and use it when we refresh Podman state after a reboot.

Fixes #2900

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
Got some lint unhappiness @mheon

**Comment by mheon** [comment]
Refresh tests are failing, which I suppose is expected given that the old locks still exist... We already deprecated `podman refresh` a few releases ago. I'm going to nuke the tests.

File: libpod/lock/shm/shm_lock.c [hunk 1]
```diff
@@ -354,6 +354,61 @@ int64_t allocate_semaphore(shm_struct_t *shm) {
   return -1 * ENOSPC;
 }
 
+// Allocate the semaphore with the given ID.
+// Returns an error if the semaphore with this ID does not exist, or has already
+// been allocated.
+// Returns 0 on success, or negative errno values on failure.
+int32_t allocate_given_semaphore(shm_struct_t *shm, uint32_t sem_index) {
```
Comment (jwhonce) [inline libpod/lock/shm/shm_lock.c]: errno's are more correctly defined as int by the standard.

File: libpod/lock/shm/shm_lock.c [hunk 2]
```diff
@@ -354,6 +354,61 @@ int64_t allocate_semaphore(shm_struct_t *shm) {
...
+  if (ret_code != 0) {
+    return -1 * ret_code;
+  }
+
+  return 0;
+}
+
 // Deallocate a given semaphore
 // Returns 0 on success, negative ERRNO values on failure
 int32_t deallocate_semaphore(shm_struct_t *shm, uint32_t sem_index) {
```
Comment (jwhonce) [inline libpod/lock/shm/shm_lock.c]: errno type is int in standard.

**Pull request merged**